pkg/collect/certificates_test.go

```diff
@@ -0,0 +1,223 @@
+6qFagl2+YQ9ng0Xwta9ezvctM3T6Dy9Kkf5OOe9ysMEsBX7s8NFxe68Qku+cExr3
+78oERlIoNOlT0cNbFLAlH2svNv1uB4qOThRDha52L+mlUdZfTMYZAwNDJWm52t/M
+NCIm5NJ5jAJpcJmoEb+JMP3j0x6wydHDXFtGm3WRggZRcrjasyodSKK6szbf96+9
+6syzAwvg9xxNtFxwbhRqqplMEz2sDWaggTrxCQzd
+-----END CERTIFICATE-----
+`
+
+var chain4 = `-----BEGIN CERTIFICATE-----
+Oy1is0whDArLNwIDAQABo4G9MIG6
+MA4GA1UdDwEB/wQEAwIE8DAdBgNVHQ4EFgQUbAavOY+vIXgc44k8GvLHH6mdzzYw
+HwYDVR0jBBgwFoAUb+S3Mu7cbwZiqZaEHTEMyUhLyH4waAYDVR0RBGEwX4IFZW52
+b3mCFGVudm95LnByb2plY3Rjb250b3VyghhlbnZveS5wcm9qZWN0Y29udG91ci5z
+dmOCJmVudm95LnByb2plY3Rjb250b3VyLnN2Yy5jbHVzdGVyLmxvY2FsMA0GCSqG
+SIb3DQEBCwUAA4IBAQBIKpBD1T9tugzJF7lajbdulXTb9qGibwQALqauskX9Sq57
+po/R2TjyxywLn4DgM7BAzzu9qfHWf+S4eQjRUHQshPbUEX9CEsSd5tCu8ZHVbBds
+6qFagl2+YQ9ng0Xwta9ezvctM3T6Dy9Kkf5OOe9ysMEsBX7s8NFxe68Qku+cExr3
+78oERlIoNOlT0cNbFLAlH2svNv1uB4qOThRDha52L+mlUdZfTMYZAwNDJWm52t/M
+NCIm5NJ5jAJpcJmoEb+JMP3j0x6wydHDXFtGm3WRggZRcrjasyodSKK6szbf96+9
+6syzAwvg9xxNtFxwbhRqqplMEz2sDWaggTrxCQzd
+-----END CERTIFICATE-----
+`
+
+// tests validate that the certParser function correctly parses a certificate
+func TestCertParser(t *testing.T) {
+
+	expiredCert := []byte(chain)
+	multiCert := []byte(chain2)
+	validCert := []byte(chain3)
+	nonCert := []byte(chain4)
+	//add docs
+
+	certParserTests := []struct {
+		Name      string
+		Certs     []byte
+		CertQty   int
+		IsSubject bool
+		IsCert    bool
+		IsValid   bool
+	}{
+		// Name, Certs, CertQty, IsSubject, IsCert, IsValid
+		{"Widgits", expiredCert, 1, true, true, false},
+		{"digicert", multiCert, 2, true, true, false}, //IsValid will not be evaluated for multicerts
+		{"envoy", validCert, 1, true, true, true},
+		{"non.crt", nonCert, 1, false, false, false}, // IsSubject and IsValid n/a for non.crt
+	}
+
+	for _, e := range certParserTests {
+
+		results, _ := CertParser(e.Name, e.Certs)
+
+		// tests if results contains a list of parsed certificates
+		var isCert bool
+		if len(results) == 0 {
+			isCert = false
+		} else {
+			isCert = true
+		}
+
+		if isCert != e.IsCert {
+			t.Errorf("Expected %v, but got %v that %s is a certificate", e.IsCert, isCert, e.Name)
+			continue
+		}
+
+		for _, cert := range results {
+
+			log.Println(e.Name, cert.Subject)
+
+			// test checks if certificate subject contains a matching string; validates that can parse cert and pull back information
+			if !strings.Contains(cert.Subject, e.Name) {
+				t.Error("unable to parse certificate: ", e.Name)
+
+			} else {
+				isSubject := true
+				if e.IsSubject != isSubject {
+					t.Errorf("You expected that %s certificate contains %s in the Subject line but it was not present; here is the entire subject string: %v", e.Name, e.Name, cert.Subject)
+				}
+			}
+
+			// test checks for expected quantity of certificates in a certificate file
+			numCerts := len(results)
+			if numCerts != e.CertQty {
+				t.Errorf("expected %d certificates in slice but got %d", e.CertQty, numCerts)
```

Comment:
we can change it to
```
assert.Equal(t, e.CertQty, len(results), "Expected %v, but got %v", e.CertQty, len(results))
```
And I think it is just one time comparing, may not need to be put into for loop, which will be the same result for each loop